test/test_vmap.py

```diff
@@ -0,0 +1,113 @@
+        x = torch.randn(2)
+        y = torch.randn(3)
+        expected_msg = 'Expected all tensors to have the same size in the mapped dimension'
+        with self.assertRaisesRegex(ValueError, expected_msg):
+            vmap(torch.mul)(x, y)
+
+    def test_func_with_no_inputs(self):
+        expected_msg = 'got no inputs'
+
+        def foo():
+            return torch.randn(3)
+
+        def bar(x):
+            return torch.randn(3)
+
+        with self.assertRaisesRegex(ValueError, expected_msg):
+            vmap(foo)()
+
+        with self.assertRaisesRegex(ValueError, expected_msg):
+            vmap(bar)()
+
+    def test_constant_function(self):
+        output = vmap(lambda x: torch.tensor(3.14))(torch.ones(3))
+        self.assertEqual(output, torch.tensor([3.14, 3.14, 3.14]))
+
+    def test_single_input(self):
+        x = torch.randn(2, 3)
+
+        def square(x):
+            return x * x
+
+        output = vmap(square)(x)
+        self.assertEqual(output, x * x)
+
+    def test_multiple_inputs(self):
+        x = torch.randn(2, 3)
+        y = torch.randn(2, 3)
+        output = vmap(torch.mul)(x, y)
+        self.assertEqual(output, x * y)
+
+    def test_multiple_outputs(self):
+        def foo(x):
+            return x * x, x * x * x
+
+        x = torch.randn(3)
+        outputs = vmap(foo)(x)
+        self.assertEqual(outputs[0], x * x)
+        self.assertEqual(outputs[1], x * x * x)
+
+    def test_nested_with_same_map_dim(self):
+        x = torch.randn(2, 3, 5)
+        y = torch.randn(2, 3, 5)
+        output = vmap(vmap(torch.mul))(x, y)
+        self.assertEqual(output, x * y)
+
+        output = vmap(vmap(vmap(torch.mul)))(x, y)
+        self.assertEqual(output, x * y)
+
+    def test_nested_with_different_map_dim(self):
+        x = torch.randn(2, 3)
+        y = torch.randn(5, 3)
+        output = vmap(lambda x: vmap(lambda y: x * y)(y))(x)
+        self.assertEqual(output, x.view(2, 1, 3) * y)
+
+        z = torch.randn(7, 3)
+        output = vmap(lambda x: vmap(lambda y: vmap(lambda z: x * y * z)(z))(y))(x)
+        self.assertEqual(output, x.view(2, 1, 1, 3) * y.view(5, 1, 3) * z)
+
+    def test_noop_in_inner_vmap(self):
+        x = torch.randn(3)
+        y = torch.randn(5)
+        output = vmap(lambda x: vmap(lambda y: x)(y))(x)
+        self.assertEqual(output, x.view(3, 1).expand(3, 5))
+
+    def test_unsupported_op_err_msg(self):
+        def foo(x):
+            return torch.cos(x)
+
+        x = torch.randn(3)
+        with self.assertRaisesRegex(RuntimeError, 'NYI: Calling aten::cos inside of vmap'):
+            vmap(foo)(x)
+
+    def test_unsupported_inplace_op_err_msg(self):
+        def foo(x):
+            return x.cos_()
+
+        x = torch.randn(3)
+        # TODO(rzou): Yeah, this error message is pretty bad because the
+        # dispatcher's fallback mechanism doesn't work for ops that don't support
+        # boxing. Fix the error message at some point.
```

Comment:
If you have ideas for how to make the default error message better, patches welcome :)